plugins/out_s3/s3.c

```diff
@@ -1242,6 +1258,193 @@ static struct multipart_upload *create_upload(struct flb_s3 *ctx,
+{
+    int ret;
+    struct upload_queue *upload_contents;
+    char *tag_cpy;
+
+    if (upload_file == NULL && m_upload_file == NULL) {
+        flb_plg_error(ctx->ins, "Tried to add entry without file to upload_queue");
+        flb_errno();
+        return -1;
+    }
+
+    if (tag_len == 0 || strlen(tag) != tag_len) {
+        flb_plg_error(ctx->ins, "Tried to add entry with invalid tag: '%s' to "
+                      "upload_queue");
+        return -1;
+    }
+
+    /* Create upload contents object and add to upload queue */
+    upload_contents = flb_malloc(sizeof(struct upload_queue));
+    if (upload_contents == NULL) {
+        flb_plg_error(ctx->ins, "Error allocating memory for upload_queue entry");
+        flb_errno();
+        return -1;
+    }
+    upload_contents->upload_file = upload_file;
+    upload_contents->m_upload_file = m_upload_file;
+    upload_contents->tag_len = tag_len;
+    upload_contents->retry_counter = 0;
+    upload_contents->upload_time = -1;
+
+    /* Necessary to create separate string for tag to prevent corruption */
+    tag_cpy = flb_malloc(tag_len);
+    if (tag_cpy == NULL) {
+        flb_plg_error(ctx->ins, "Error allocating memory for tag in add_to_queue");
+        flb_errno();
+        return -1;
+    }
+    strncpy(tag_cpy, tag, tag_len);
+    upload_contents->tag = tag_cpy;
+
+    /* Adding to mk_list is not thread safe, so mutex lock is necessary */
+    ret = pthread_mutex_trylock(&ctx->upload_queue_mutex);
+    if (ret != 0) {
+        flb_plg_debug(ctx->ins, "Failed to lock mutex in add_to_queue");
+        return -1;
+    }
+    mk_list_add(&upload_contents->_head, &ctx->upload_queue);
+    ret = pthread_mutex_unlock(&ctx->upload_queue_mutex);
+    if (ret) {
+        flb_plg_error(ctx->ins, "Failed to unlock mutex for add_to_queue");
+        flb_errno();
+        return -1;
+    }
+
+    return 0;
+}
+
+/* Removes an entry from upload_queue */
+void remove_from_queue(struct upload_queue *entry) {
+    mk_list_del(&entry->_head);
+    flb_free(entry->tag);
+    flb_free(entry);
+    return;
+}
+
+/* Uploads all chunk files in queue synchronously */
+static void s3_upload_queue(struct flb_config *config, void *out_context)
+{
+    int ret;
+    int async_flags;
+    time_t now;
+    char *buffer;
+    size_t buffer_size;
+    struct upload_queue *upload_contents;
+    struct flb_s3 *ctx = out_context;
+    struct mk_list *tmp;
+    struct mk_list *head;
+
+    flb_plg_debug(ctx->ins, "Running upload timer callback..");
+
+    /* mutex lock */
+    ret = pthread_mutex_trylock(&ctx->upload_queue_mutex);
+    if (ret != 0) {
+        flb_plg_error(ctx->ins, "Failed to lock mutex in upload_queue");
+        return;
+    }
+
+    /* upload timer must use sync mode */
+    if (ctx->use_put_object == FLB_TRUE) {
+        async_flags = ctx->s3_client->upstream->flags;
+        ctx->s3_client->upstream->flags &= ~(FLB_IO_ASYNC);
```

Comment:
I believe async flag is defined somewhere else and this is the safest way to do this, though I am not too sure. @PettitWesley would know better about this as I was copying the logic used in other functions.

Reply:
Yeah he just copied this code from me. I think this code is best since it works no matter what all flags are set. Multiple flags can be set at one time.